lightly/models/modules/masked_causal_vision_transformer.py

```diff
@@ -0,0 +1,247 @@
+        dynamic_img_size: bool = False,
+        dynamic_img_pad: bool = False,
+        drop_rate: float = 0.0,
+        pos_drop_rate: float = 0.0,
+        patch_drop_rate: float = 0.0,
+        proj_drop_rate: float = 0.0,
+        attn_drop_rate: float = 0.0,
+        drop_path_rate: float = 0.0,
+        weight_init: Literal["skip", "jax", "jax_nlhb", "moco", ""] = "",
+        embed_layer: Callable = PatchEmbed,  # type: ignore[type-arg]
+        norm_layer: Optional[LayerType] = None,
+        act_layer: Optional[LayerType] = None,
+        block_fn: Type[Module] = MaskedCausalBlock,
+        mlp_layer: Type[Module] = Mlp,
+    ) -> None:
+        super().__init__(
+            img_size=img_size,
+            patch_size=patch_size,
+            in_chans=in_chans,
+            num_classes=num_classes,
+            global_pool=global_pool,
+            embed_dim=embed_dim,
+            depth=depth,
+            num_heads=num_heads,
+            mlp_ratio=mlp_ratio,
+            qkv_bias=qkv_bias,
+            qk_norm=qk_norm,
+            init_values=init_values,
+            class_token=class_token,
+            no_embed_class=no_embed_class,
+            reg_tokens=reg_tokens,
+            pre_norm=pre_norm,
+            fc_norm=fc_norm,
+            dynamic_img_size=dynamic_img_size,
+            dynamic_img_pad=dynamic_img_pad,
+            drop_rate=drop_rate,
+            pos_drop_rate=pos_drop_rate,
+            patch_drop_rate=patch_drop_rate,
+            proj_drop_rate=proj_drop_rate,
+            attn_drop_rate=attn_drop_rate,
+            drop_path_rate=drop_path_rate,
+            weight_init=weight_init,
+            embed_layer=embed_layer,
+            norm_layer=norm_layer,
+            act_layer=act_layer,
+            block_fn=block_fn,
+            mlp_layer=mlp_layer,
+        )
+
+    def forward_features(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
+        """Forward pass of the model without the classification head.
+
+        Args:
+            x:
+                Input tensor of shape (batch_size, sequence_length, channels).
+            mask:
+                Mask of shape (batch_size, sequence_length) indicating which tokens
+                should be masked. Tokens where the mask is True will only be used for
+                causal attention, while unmasked tokens are used for bidirectional
+                attention. If the mask is None, all tokens are used for bidirectional
+                attention.
+        """
+        x = self.patch_embed(x)
+        x = self._pos_embed(x)
+        x = self.patch_drop(x)
+        x = self.norm_pre(x)
+        if self.grad_checkpointing and not jit.is_scripting():
+            # TODO: This probably doesn't work correctly as it doesn't consider the
+            # mask.
+            x = _manipulate.checkpoint_seq(self.blocks, x)
+        else:
+            for block in self.blocks:
+                x = block(x, mask=mask)
```

Comment:
I assume that you have to loop through each `block` here because of the `mask`?

Reply:
Yes, `self.blocks` is a `Sequential` object which only take a single input `x` in the forward method: https://pytorch.org/docs/stable/_modules/torch/nn/modules/container.html#Sequential We have to pass the mask manually :)